postgres-persistence/src/main/java/com/netflix/conductor/postgres/util/PostgresQueueListener.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2024 Conductor Authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.netflix.conductor.postgres.util;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Iterator;
+
+import javax.sql.DataSource;
+
+import org.postgresql.PGConnection;
+import org.postgresql.PGNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.conductor.core.exception.NonTransientException;
+import com.netflix.conductor.postgres.config.PostgresProperties;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class PostgresQueueListener {
+
+    private PGConnection pgconn;
+
+    private Connection conn;
+
+    private DataSource dataSource;
+
+    private HashMap<String, QueueStats> queues;
+
+    private boolean connected = false;
+
+    private boolean connecting = false;
+
+    private long lastNotificationTime = 0;
+
+    private Integer stalePeriod;
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public PostgresQueueListener(DataSource dataSource, PostgresProperties properties) {
+        logger.info("Using experimental PostgresQueueListener");
+        this.dataSource = dataSource;
+        this.stalePeriod = properties.getExperimentalQueueNotifyStalePeriod();
+        connect();
+    }
+
+    public boolean hasMessagesReady(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return true;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
+        if (queueStats == null) {
+            return false;
+        }
+
+        if (queueStats.getNextDelivery() > System.currentTimeMillis()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public Integer getSize(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return null;
+        }
```

Comment:
Nit: prefer returning optional here, implicit nulls, while handled well now may result in bugs later.